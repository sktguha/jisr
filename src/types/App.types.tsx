export interface folderData {
  id: string;
  length?: number;
  name?: string;
  isfolder?: boolean;
  children: folderData[];
}

export type genericFn = (...args: any[])=>any

