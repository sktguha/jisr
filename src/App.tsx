import React, { useEffect, useState } from 'react';
import './style.css';

import { initialFolderData } from './data/folderData';
import FolderNew from './components/FolderNew';

import useFolder from './hooks/useFolder';
import { folderData } from './types/App.types';

function flatten(treeset: folderData, arr: any[]): any[] {
  arr.push(treeset);
  if (treeset.children) {
    treeset.children.forEach((elem) => {
      flatten(elem, arr);
    })
  }
  return arr;
}

function SearchBar({ treeset }: { treeset: folderData }) {
  const onChange = (e: { target: { value: any; }; }) => {
    // @ts-ignore
    [...document.querySelectorAll('.search-bar-active')].forEach((dom) => dom.classList.remove('search-bar-active'));
    const value = e.target.value;
    if (!value) return;
    const elems = flatten(treeset, []);
    elems.forEach((elem) => {
      const dom = document.getElementById(elem.id);
      if (!dom) return;
      if (elem.name.indexOf(value) !== -1) {
        dom.classList.add('search-bar-active');
      }
    })
  }
  return <div>
    <h3 className='inline'>Search bar</h3> <input className="search-input" type="text" onChange={onChange}></input>
  </div>
}

export default function App() {
  const [treeset, setTreeset] = useState(initialFolderData as unknown as folderData | null);
  const { addnewItem, updateItem, deleteItem } = useFolder();
  const addItem = (id: string, name: string, isfolder: boolean) => {
    setTreeset(addnewItem(treeset as folderData, id, name, isfolder));
  };

  const rename = (id: string, name: string) => {
    setTreeset(updateItem(treeset as folderData, id, name));
  };

  useEffect(() => {
    document.body.onkeydown = (el: any) => {
      if (el.keyCode === 40 || el.keyCode === 38) {
        el.preventDefault();
        let isDown = false;
        if (el.keyCode === 38) { isDown = true; }
        const elems = flatten(treeset as folderData, []);
        const currDom = document.querySelector('.key-active');
        let currId;
        if (currDom) {
          currDom.classList.remove('key-active');
          currId = currDom?.id;
        }
        else {
          currId = elems[isDown ? elems.length - 1 : 0].id;
          const dom = document.getElementById(currId) as HTMLElement;
          dom?.classList.add('key-active');
          dom.tabIndex = -1;
          dom?.focus();
          return;
        }
        let i;
        for (i = 0; i < elems.length; i++) {
          if (elems[i].id === currId) break;
        }
        if (!isDown) {
          for (let j = i + 1; j < elems.length; j++) {
            const dom = document.getElementById(elems[j].id);
            if (dom) {
              currId = elems[j].id;
              dom.classList.add('key-active');
              dom.tabIndex = -1;
              dom.focus();
              break;
            }
          }
        } else {
          for (let j = i - 1; j > -1; j--) {
            const dom = document.getElementById(elems[j].id);
            if (dom) {
              currId = elems[j].id;
              dom.classList.add('key-active');
              dom.tabIndex = -1;
              dom.focus();
              break;
            }
          }
        }
        return false;
      }
    }
  }, [treeset]);


  const delete_item = (id: string) => {
    // do not delete the root item
    if (id === initialFolderData.id) return;

    setTreeset(deleteItem(treeset as folderData, id));
  };

  return (
    <div className="main-container">
      <h2 className="title-header" >FILE MANAGER</h2>
      <SearchBar treeset={treeset as folderData} />
      <p>(Press Enter to submit the input value)</p>
      <FolderNew
        treeset={treeset as folderData}
        addItem={addItem}
        rename={rename}
        delete_item={delete_item}
      />
    </div>
  );
}
