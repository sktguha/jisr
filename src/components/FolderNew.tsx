import React, { useState } from 'react';
import '../style.css';
import { folderData, genericFn } from '../types/App.types';
import { RightMenu } from './RightMenu';

export default function FolderNew({ treeset, addItem, rename, delete_item }: { treeset: folderData, addItem: genericFn, rename: genericFn, delete_item: genericFn }) {
    const [expand, setExpand] = useState(true); // can set to false to keep collapsed by default
    const operationType = React.useRef('new');
    const [minput, setInput] = useState({
        value: '',
        show: false,
        placeholder: '',
    });

    const [isRightMenu, setIsRightMenu] = useState(false);
    const onKeyDown = (e: { keyCode: number; }) => {
        if (e.keyCode === 67) {
            const currDom = document.querySelector('.key-active');
            let keyId;
            if (currDom) {
                keyId = currDom.id;
            } else {
                keyId = 0;
            }
            if (keyId === treeset.id) {
                setIsRightMenu(true);
            }
        }
    };
    if (treeset.isfolder) {
        return (
            <div className="folder-item" key={treeset.id}>
                <div className="folder"
                    onContextMenu={(e: any) => {
                        setIsRightMenu(true);
                        e.preventDefault();
                        return false;
                    }}>
                    <span onClick={() => setExpand(!expand)}>{expand ? '⏬' : '⏩'}</span>
                    <div className="name-disp" id={treeset.id} onKeyDown={onKeyDown}>📁{treeset.name}</div>
                    {minput.show && (
                        <input
                            placeholder={minput.placeholder}
                            onKeyDown={(el) => {
                                //@ts-ignore
                                if (el.keyCode === 13 && el.target.value) {
                                    if (operationType.current === 'new') {
                                        minput.placeholder.toString().includes('folder')
                                            ? addItem(treeset.id, minput.value, true)
                                            : addItem(treeset.id, minput.value, false);
                                    } else if (operationType.current === 'update') {
                                        rename(treeset.id, minput.value);
                                    }
                                    setInput({
                                        show: false,
                                        value: '',
                                        placeholder: '',
                                    });
                                }
                            }}
                            autoFocus
                            value={minput.value}
                            onChange={(e) => setInput({ ...minput, value: e.target.value })}
                        />
                    )}
                    {isRightMenu && <RightMenu
                        fileName={treeset.name}
                        setIsRightMenu={setIsRightMenu}
                        onAddNewFolder={(el) => {
                            setInput((e) => {
                                return {
                                    ...e,
                                    show: true,
                                    placeholder: 'enter folder name',
                                };
                            });
                            operationType.current = 'new';
                        }}
                        onAddNewFile={(el) => {
                            setInput({
                                ...minput,
                                show: true,
                                placeholder: 'enter file name',
                            });
                            operationType.current = 'new';
                            setIsRightMenu(false);
                        }}
                        onRename={() => {
                            setInput({
                                ...minput,
                                show: true,
                                placeholder: 'enter new name',
                            });
                            operationType.current = 'update';
                        }}
                        onDelete={(el) => {
                            delete_item(treeset.id);
                        }}
                    />}
                </div>
                {expand &&
                    treeset.children.map((e) => {
                        return (
                            <FolderNew
                                key={e.id}
                                treeset={e}
                                addItem={addItem}
                                rename={rename}
                                delete_item={delete_item}
                            />
                        );
                    })}
            </div>
        );
    } else {
        return (
            <div className="file file-item" key={treeset.id}
                onContextMenu={(e: any) => {
                    setIsRightMenu(true);
                    e.preventDefault();
                    return false;
                }}
            >
                <div id={treeset.id} className="name-disp" tabIndex={-1}
                    onKeyDown={onKeyDown}
                >{(() => {
                    if (treeset.name?.endsWith('.png') || treeset.name?.endsWith('.jpeg') || treeset.name?.endsWith('.svg')
                        || treeset.name?.endsWith('.gif') || treeset.name?.endsWith('.webp')) {
                        return '📷';
                    } else if (treeset.name?.endsWith('css') || treeset.name?.endsWith('js') || treeset.name?.endsWith('ts') || treeset.name?.endsWith('html')) {
                        return '🖥';
                    } else {
                        return '📄';
                    }
                })()}{treeset.name}</div>
                {minput.show && (
                    <input
                        placeholder={minput.placeholder}
                        onKeyDown={(el) => {
                            //@ts-ignore
                            if (el.keyCode === 13 && el.target.value) {
                                if (operationType.current === 'new') {
                                    minput.placeholder.toString().includes('folder')
                                        ? addItem(treeset.id, minput.value, true)
                                        : addItem(treeset.id, minput.value, false);
                                } else if (operationType.current === 'update') {
                                    rename(treeset.id, minput.value);
                                }
                                setInput({
                                    show: false,
                                    value: '',
                                    placeholder: '',
                                });
                            }
                        }}
                        autoFocus
                        value={minput.value}
                        onChange={(e) => setInput({ ...minput, value: e.target.value })}
                    />
                )}
                {isRightMenu && <RightMenu
                    isFile={true}
                    setIsRightMenu={setIsRightMenu}
                    fileName={treeset.name}
                    onRename={() => {
                        console.log('clicked');
                        setInput({
                            ...minput,
                            show: true,
                            placeholder: 'enter new name',
                        });
                        operationType.current = 'update';
                    }}
                    onDelete={() => {
                        delete_item(treeset.id);
                    }}
                />}
            </div>
        );
    }
}
