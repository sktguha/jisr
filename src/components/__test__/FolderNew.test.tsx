import React from 'react';
import { render, screen } from '@testing-library/react';
import FolderNew from '../FolderNew';
import { folderData } from '../../types/App.types';

const emptyFn = () => { };
const Fns = {
    addItem: emptyFn,
    rename: emptyFn,
    delete_item: emptyFn
}

test('renders FolderNew folder', () => {
    const treeset = {
        id: '1',
        name: 'test',
        isfolder: true,
        children: []
    }
    const { container } = render(<FolderNew treeset={treeset as folderData} {...Fns} />)
    expect(container).toMatchSnapshot()
});

test('renders FolderNew file images', () => {
    const treeset = {
        id: '1',
        name: 'test',
        isfolder: true,
        children: [{
            id: '2',
            name: 'image.webp'
        }
        ]
    }
    const { container } = render(<FolderNew treeset={treeset as folderData} {...Fns} />)
    expect(container).toMatchSnapshot()
});

test('renders FolderNew file code', () => {
    const treeset = {
        id: '1',
        name: 'test',
        isfolder: true,
        children: [
            {
                id: '2',
                name: 'style.css'
            }
        ]
    }
    const { container } = render(<FolderNew treeset={treeset as folderData} {...Fns} />)
    expect(container).toMatchSnapshot()
});

test('renders FolderNew file others', () => {
    const treeset = {
        id: '1',
        name: 'test',
        isfolder: true,
        children: [{
            id: '2',
            name: 'notes.txt'
        }]
    }
    const { container } = render(<FolderNew treeset={treeset as folderData} {...Fns} />)
    expect(container).toMatchSnapshot()
});
