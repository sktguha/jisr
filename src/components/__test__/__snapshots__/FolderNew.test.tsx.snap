// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FolderNew file code 1`] = `
<div>
  <div
    class="folder-item"
  >
    <div
      class="folder"
    >
      <span>
        ⏬
      </span>
      <div
        class="name-disp"
        id="1"
      >
        📁
        test
      </div>
    </div>
    <div
      class="file file-item"
    >
      <div
        class="name-disp"
        id="2"
        tabindex="-1"
      >
        🖥
        style.css
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FolderNew file images 1`] = `
<div>
  <div
    class="folder-item"
  >
    <div
      class="folder"
    >
      <span>
        ⏬
      </span>
      <div
        class="name-disp"
        id="1"
      >
        📁
        test
      </div>
    </div>
    <div
      class="file file-item"
    >
      <div
        class="name-disp"
        id="2"
        tabindex="-1"
      >
        📷
        image.webp
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FolderNew file others 1`] = `
<div>
  <div
    class="folder-item"
  >
    <div
      class="folder"
    >
      <span>
        ⏬
      </span>
      <div
        class="name-disp"
        id="1"
      >
        📁
        test
      </div>
    </div>
    <div
      class="file file-item"
    >
      <div
        class="name-disp"
        id="2"
        tabindex="-1"
      >
        📄
        notes.txt
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FolderNew folder 1`] = `
<div>
  <div
    class="folder-item"
  >
    <div
      class="folder"
    >
      <span>
        ⏬
      </span>
      <div
        class="name-disp"
        id="1"
      >
        📁
        test
      </div>
    </div>
  </div>
</div>
`;
